### PR TITLE
Improve internal mechanisms of the nitvm

### PR DESCRIPTION
**Pull request by jpages**
This PR fixes a bug that was hidden in the nitvm for a too long time, the recursion in class loading was bad and is now corrected.

Then a more accurate mechanism is introduce to deal with positions of methods and attributes in internal structures of classes and objects.
When a method dispatch is performed, this mechanism allows to ask precisely for a class the position of a block of methods (or attributes). This should increase the proportion of object mechanisms "compiled" in an efficient way.

Finally, the class loading is optimized: an indirectly loaded class is not fully constructed and allocated.
Instead, its identifier is computed to allow subtyping tests.

For now, these more accurate mechanisms need recompilations of callsite. This sould be done in further PR.

### CODE REVIEW

File: src/vm/virtual_machine.nit [hunk 1]
```diff
@@ -376,18 +405,33 @@ redef class MClass
...
+	# The position of the class' block in virtual table,
+	# the position is set to -1 when the invariant position is no longer satisfied
+	var position_attributes: Int
+
+	# The position of the class' block in attribute table
+	# the position is set to -1 when the invariant position is no longer satisfied
+	var position_methods: Int
+
+	# The chosen prefix for this class
+	var prefix: nullable MClass
```
Comment (privat) [inline src/vm/virtual_machine.nit]: wat?

File: src/vm/virtual_machine.nit [hunk 2]
```diff
@@ -401,15 +445,14 @@ redef class MClass
 	var mmethods = new Array[MMethod]
 
 	# Allocates a VTable for this class and gives it an id
-	private fun make_vt(v: VirtualMachine)
+	# * `vm` The currently executed VirtualMachine
+	# * `explicit` Indicate if this class was explicitly loaded
```
Comment (privat) [inline src/vm/virtual_machine.nit]: what is the effect of `explicit`?

File: src/vm/virtual_machine.nit [hunk 3]
```diff
@@ -818,4 +1001,9 @@ class MemoryManager
 			MMethodDef_incr_ref(method);
 		}
 	`}
+
+	# Return a NULL pointer, used to initialize virtual tables
+	private fun null_ptr: Pointer `{
+		return NULL;
+	`}
```
Comment (privat) [inline src/vm/virtual_machine.nit]: @xymus why this is not in the FFI?

**Comment by privat** [comment]
+1 for the rest

**Comment by jpages** [comment]
Reroll to improve the doc.

For your remark, I needed to instantiate a null Pointer to initialize an attribute, maybe this could be add to the Pointer class ?

**Comment by R4PaSs** [comment]
:shipit:

**Pull request merged**